File: Python/Binary Search Tree/Delete_Node.py
# Delete a Node from a Binary Search Tree
# Time Complexity: O(logN) and Space complexity: O(Recursion stack size)
# author: Pavan Kumar Paluri

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        # base case 
        if root is None:
            return None 
        def successor(current: TreeNode):
            # successor: one right go until last left 
            current = current.right 
            while current.left is not None:
                current = current.left 
            return current.val
        def predecessor(current: TreeNode):
            # predecessor: One left and go until last right node
            current = current.left 
            while current.right is not None:
                current = current.right
            return current.val
        
        # if given key is > root.val -> proceed right 
        if key > root.val:
            root.right = self.deleteNode(root.right, key)
        elif key < root.val:
            # explore left subtree
            root.left = self.deleteNode(root.left, key)
        else:
            # if here: Node itself is the key - to be deleted
            # case-1 : if node is a leaf node 
            if root.left is None and root.right is None:
                # delete the node
                root = None
            # case-2: If right node is not None
            elif root.right is not None:
                root.val = successor(root)
                # delete all the nodes 
                root.right = self.deleteNode(root.right, root.val)
            elif root.right is None:
                root.val = predecessor(root)
                # delete all the nodes 
                root.left = self.deleteNode(root.left, root.val)
        return root

File: C++/Binary Search Tree/Delete_Node.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // inorder traversal - has all sorted in ascending order
    int left_pred(TreeNode* root)
    {
        // take one right and proceed until last left node
        root = root->left;
        while(root->right != nullptr)
        {
            // proceed until last left node
            root = root->right;
        }
        // if here: the node is the leftmost node
        return root->val;
    }
    
    int right_succ(TreeNode* root)
    {
        root = root->right;
        while(root->left != nullptr)
        {
            root = root->left;
        }
        return root->val;
    }
    
    TreeNode* deleteNode(TreeNode* root, int key) {
        // if root is Null: return None
        if(root == nullptr)
            return nullptr;
        // if given key is greater than root node in a BST
        if(key > root->val)
        {
            // explore right subtree 
            root->right = deleteNode(root->right, key);
        }
        else if(key < root->val)
        {
            // explore left subtree
            root->left = deleteNode(root->left, key);
        }
        else{
            // if here: key == root->val and is a leaf node 
            if(key == root->val && root->left == nullptr && root->right == nullptr) // this is the desired node to be deleted
                root = nullptr;
            // if there exists a right subtree: explore that
            else if(root->right != nullptr)
            {
                root->val = right_succ(root);
                // delete that node after making current node the val
                root->right = deleteNode(root->right, root->val);
            }
            else {
                root->val = left_pred(root);
                root->left = deleteNode(root->left, root->val);
            }
        }
        return root;
    }
};
